hw/opentitan/ot_usbdev.c

```diff
@@ -0,0 +1,827 @@
+    }
+}
+
+static void ot_usbdev_update_alerts(OtUsbdevState *s)
+{
+    uint32_t level = s->regs[R_ALERT_TEST];
+    s->regs[R_ALERT_TEST] = 0u;
+
+    /* @todo add other sources of alerts here */
+
+    if (level & R_ALERT_TEST_FATAL_FAULT_MASK) {
+        ibex_irq_set(&s->alert, 1);
+        ibex_irq_set(&s->alert, 0);
+    }
+}
+
+/*
+ * Register read/write handling
+ */
+
+static uint64_t ot_usbdev_read(void *opaque, hwaddr addr, unsigned size)
+{
+    OtUsbdevState *s = opaque;
+    (void)size;
+    uint32_t val32;
+
+    hwaddr reg = R32_OFF(addr);
+    switch (reg) {
+    case R_USBDEV_INTR_STATE:
+    case R_USBDEV_INTR_ENABLE:
+    case R_USBDEV_INTR_TEST:
+        val32 = s->regs[reg];
+        break;
+    case R_ALERT_TEST:
+        qemu_log_mask(LOG_GUEST_ERROR,
+                      "%s: %s Read to W/O register 0x%02" HWADDR_PRIx " (%s)\n",
+                      __func__, s->ot_id, addr, REG_NAME(reg));
+        val32 = 0;
+        break;
+    case R_USBCTRL:
+    case R_EP_OUT_ENABLE:
+    case R_EP_IN_ENABLE:
+    case R_USBSTAT:
+        /*
+         * For now, there is no host so we always always operate in
+         * vbus-override mode.
+         * @todo Report other statuses.
+         */
+        val32 = s->link_state << R_USBSTAT_LINK_STATE_SHIFT |
+                (uint32_t)s->vbus_sense << R_USBSTAT_SENSE_SHIFT;
+        break;
+    case R_AVOUTBUFFER:
+    case R_AVSETUPBUFFER:
+    case R_RXFIFO:
+    case R_RXENABLE_SETUP:
+    case R_RXENABLE_OUT:
+    case R_SET_NAK_OUT:
+    case R_IN_SENT:
+    case R_OUT_STALL:
+    case R_IN_STALL:
+    case R_CONFIGIN_0:
+    case R_CONFIGIN_1:
+    case R_CONFIGIN_2:
+    case R_CONFIGIN_3:
+    case R_CONFIGIN_4:
+    case R_CONFIGIN_5:
+    case R_CONFIGIN_6:
+    case R_CONFIGIN_7:
+    case R_CONFIGIN_8:
+    case R_CONFIGIN_9:
+    case R_CONFIGIN_10:
+    case R_CONFIGIN_11:
+    case R_OUT_ISO:
+    case R_IN_ISO:
+    case R_OUT_DATA_TOGGLE:
+    case R_IN_DATA_TOGGLE:
+    case R_PHY_PINS_SENSE:
+    case R_PHY_PINS_DRIVE:
+    case R_PHY_CONFIG:
+    case R_WAKE_CONTROL:
+    case R_WAKE_EVENTS:
+    case R_FIFO_CTRL:
+    case R_COUNT_OUT:
+    case R_COUNT_IN:
+    case R_COUNT_NODATA_IN:
+    case R_COUNT_ERRORS:
+        qemu_log_mask(LOG_UNIMP, "%s: %s %s is not supported\n", __func__,
+                      s->ot_id, REG_NAME(reg));
+        val32 = 0;
+        break;
+    default:
+        qemu_log_mask(LOG_GUEST_ERROR, "%s: %s Bad offset 0x%" HWADDR_PRIx "\n",
+                      __func__, s->ot_id, addr);
+        val32 = 0;
+        break;
+    }
+
+    uint32_t pc = ibex_get_current_pc();
+    trace_ot_usbdev_io_read_out(s->ot_id, (uint32_t)addr, REG_NAME(reg), val32,
+                                pc);
+
+    return (uint64_t)val32;
+}
+
+static void ot_usbdev_write(void *opaque, hwaddr addr, uint64_t val64,
+                            unsigned size)
+{
+    OtUsbdevState *s = opaque;
+    (void)size;
+    uint32_t val32 = val64;
```

Comment:
Should probably be an explicit cast.
```suggestion
    uint32_t val32 = (uint32_t)val64;
```